### PR TITLE
ui: dashboard links

### PR DESCRIPTION
**Pull request by johncowen**
This PR adds a new `{{template-anchor}}` component. This component lets you specify a 'href template' in a handlebars like format instead of a normal string href. This template will be interpolated with the contents of a `vars=""` attribute.

~`href`s are currently purposefully left blank.~

~`href`s use the `consul-dashboard-url` meta value.~

`href`s use a UI Setting that you can add on a per user basis via the UI Settings page (see grabs below)

Screens currently use the current styling/widgets - these will be updated across the UI as a whole 🔜 

Settings page:

Screenshot 2019-05-01 at 12 55 51

Dashboard link:

Screenshot 2019-05-01 at 12 56 26

### CODE REVIEW

File: ui-v2/app/components/templated-anchor.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        wm.set(ref, map);
+      }
+      map[prop] = value;
+      return map[prop];
+    },
+  };
+};
+const weak = createWeak();
+
+const templateRe = /{{([A-Za-z.]*)}}/g;
```
Comment (banks) [inline ui-v2/app/components/templated-anchor.js]: Do you need to escape `{` and `}` in JS regexp? I guess not if this works just wondering if it's one of those things where not escaping seems to work but accidentally matches other stuff like how `foo.com` matches `fooZcom`? Other examples I've seen of handlebars syntax regexp in JS do escape the opening ones at least?

File: ui-v2/tests/integration/components/templated-anchor-test.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    {{/templated-anchor}}
+  `);
+
+  assert.equal(
+    this.$()
+      .text()
+      .trim(),
+    'template block text'
+  );
+});
```
Comment (banks) [inline ui-v2/tests/integration/components/templated-anchor-test.js]: I could be missing something but it looks like this doesn't actually test the variable replacement? Do we know that regexp etc. works?

File: ui-v2/app/components/templated-anchor.js (hunk 1 above)
Comment (johncowen) [reply ui-v2/app/components/templated-anchor.js]: Not that I'm aware of, it's potentially one of those 'optional' escapes that eslint warns you that it's not needed, will take a look though. First glance its looks like `\{` and `{` both match the same thing.

Comment (johncowen) [reply ui-v2/app/components/templated-anchor.js]: Ok so, if I've understood correctly, generally you do not need to escape `{`s unless it's part of a repetition operator (like `{1,3}`) - but you can if you want. I think I'll put some comments above this at least though. I usually do that with RegExp to make them super clear.

Actually, the more I look at this the more I wonder, I should probably switch the `*` to a `?`. Plus, do you know whether we are allowing people to use meta data for interpolation, I was trying to keep the matches pretty restrictive just for our API keys but maybe I should add `0-9` and `_-` in there?

> Key can contain only ASCII chars and no special characters (A-Z a-z 0-9 _ and -)

What do you think?

File: ui-v2/tests/integration/components/templated-anchor-test.js (hunk 2 above)
Comment (johncowen) [reply ui-v2/tests/integration/components/templated-anchor-test.js]: So these are just the auto generated tests when you use an ember generator to create your stubs. They generally just test the thing you generate can be instantiated.

I've added some integration tests with some basic usecases which cover the regex.

**Comment by johncowen** [comment]
Hey! So I remembered the talk we had last week where we discussed this, and I vaguely remember we briefly discussed being able to use meta keys for interpolation here, so I added those extra char in as well as adding some basic testing around usage.

File: ui-v2/app/components/templated-anchor.js (hunk 1 above)
Comment (banks) [reply ui-v2/app/components/templated-anchor.js]: Restrictive is great... see reply in main conversation.

**Comment by banks** [comment]
Thanks @johncowen,

> and I vaguely remember we briefly discussed being able to use meta keys for interpolation here

I guess this highlights the risk of working on even simple features that weren't well documented just discussed verbally.

The intent of this feature is that we _only_ support interpolating a very small handfull of values from the current context in the UI:

 - `SERVICE_NAME` - the name of the service being viewed (e.g. on service listing pages AND service instance pages)
 - `SERVICE_ID` - the ID of the service _instance_ being viewed (only on service instance pages)
 - `NODE` - the ID(name) of the node on which the service _instance_ is (service instance pages AND node pages)

The whole idea of templating these on the front end came from the original plan that these would be configured centrally (i.e. UI fetches them separately from the service instances themselves).

In moving to pulling the url template from service meta, it basically doesn't make any sense to have this templating done in the UI any more... Consul servers also have all the context the UI does at this point and could just resolve the template themselves in the API which is way more useful.

I'm going to think about this a little more and we can talk about it tomorrow.

**Comment by johncowen** [comment]
Hey Paul!

Thanks for reviewing this, looks like you caught a biggie there! 😅 

Yeah for sure, with the talk of being able to interpolate meta data and the suggestion of `consul-dashboard-tpl` for the name of the meta key - what we have here was what seemed to make sense 🤷‍♂ 

Don't worry though, if you are going to do all the above in the backend (template interpolation etc), the above PR should still 'work fine' (and I learnt why I don't have to escape `{`s!). I'm glad we eventually chose the `consul-dashboard-url` name for the key instead, it means this still 'makes sense'. I should probably remove the extra interpolation here though. Although I'll wait until something is written down to that effect first, so I don't repeat the same mistake.

> The intent of this feature is that we only support interpolating a very small handfull of values from the current context in the UI:

> SERVICE_NAME - the name of the service being viewed (e.g. on service listing pages AND service instance pages)
> SERVICE_ID - the ID of the service instance being viewed (only on service instance pages)
> NODE - the ID(name) of the node on which the service instance is (service instance pages AND node pages)

I'm not sure if you are still going with the plan of these strings for interpolation (do we use these values anywhere else), but if you are set on these, then we should probably at least think about making the `NODE` one into `NODE_ID/_NAME` (nodes have `Node` and `ID` properties, it could probably be confusing if we just used the word `NODE`. I'm confused now even! ("NODE - the ID(name)" > do you mean the ID or the name? We use `Node` to refer to the name in the API but you also mention `ID` here which is a different property in the API) What do you think?

> I'm going to think about this a little more and we can talk about it tomorrow.

Cool, gimme a yell when you are about, but we should probably get something written down that you are happy with first so we aren't just discussing these things verbally again!

**Comment by johncowen** [comment]
This is ready to go again, I also added some screengrabs to the PR description

**Comment by pearkes** [comment]
Looks like Ember tests are failing here ^

**Comment by johncowen** [comment]
Cool, missed a little dependency for a unit test, thanks for spotting! Test are passing again now

**Pull request merged**